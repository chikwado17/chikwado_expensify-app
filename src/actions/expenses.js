import database from '../firebase/firebase';


//creating action generator for addExpense
                                //creating default object for expenses then pass it to expenses object                        
export const addExpense = (expense) => ({
    type: 'ADD_EXPENSE',
    expense
});

//action dispatch of firebasre to redux 
export const startAddExpense = (expenseData = {}) => {
   
    return (dispatch) => {

        const {
            description = '',
            note = '',
            amount = 0,
            createdAt = 0
        } = expenseData;


        database.ref('expenses').push({ description,note,amount,createdAt }).then((ref)=>{
            //dispatching the function generator for database
            dispatch(addExpense({
                id:ref.key,   
                description,
                note,
                amount,
                createdAt
            }));
        });
    };
};



//creating action generator for removeExpense
export const removeExpense = ({id} = {}) => ({
    type: 'REMOVE_EXPENSE',
    id
});

//action generator for editExpense
export const editExpense = (id,updates) => ({
    type:'EDIT_EXPENSE',
    id,
    updates
});
//end of expenses reducer



//setup for firebase to fetch save data from database

//set_Expenses action generator
export const setExpenses = (expenses) => ({
    type:'SET_EXPENSES',
    expenses
});

//dispatching function for setExpenses for firebase database
export const startSetExpenses = () => {
    return (dispatch) => {
       return database.ref('expenses').once('value').then((snapshot) => {
            const expenses = [];

            snapshot.forEach((childSnapshot) => {
                    expenses.push({
                        id:childSnapshot.key,
                        ...childSnapshot.val()
                    });
            });

            dispatch(setExpenses(expenses));
        });
    }
}

