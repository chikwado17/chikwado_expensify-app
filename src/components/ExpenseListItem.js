import React, { Component } from 'react';
import { Link } from 'react-router-dom';
import moment from 'moment';
import numeral from 'numeral';

                            //destructing ID and Dispatch as a props
const ExpenseListItem = ({ id, description,amount,createdAt}) => {
    return ( 
        <div>
            {/* Redirecting the links to edit page with the props ID*/}
           <Link to={`/edit/${id}`}>
                 <h3>{description}</h3>
           </Link>
           <p>
                {numeral(amount / 100).format('$0,0.00')} 
                -
                {moment(createdAt).format('MMMM Do, YYYY')}
            </p>
        </div>
     );
}

export default ExpenseListItem;