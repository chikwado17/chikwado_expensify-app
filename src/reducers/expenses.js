// creating a default value for expensesReducer state
const expensesReducerDefault = [];
//creating my expenseReducer
const expensesReducer = (state = expensesReducerDefault,action) => {
    switch(action.type){
        case 'ADD_EXPENSE':
        //using spread operator to add the current expense value to the old expense value without changing the expense object
            return[
                ...state,
                action.expense
            ];
        case 'REMOVE_EXPENSE':
            //removing the expenseOne object your the id, if the id is equal to the expense
            return state.filter(({id}) =>  id !== action.id);
            //updating the match with is the amount using map function to check through the object
        case 'EDIT_EXPENSE':
            return state.map((expense) => {
                if(expense.id === action.id){
                    return {
                    ...expense,
                        //finding the update from action.update which is coming from the exitExpense action generator
                    ...action.updates
                    };
                }else{
                    return expense;
                };
            }); 
        case 'SET_EXPENSES':
            return action.expenses
        default:
            return state;
    }
}
//End of expenseReducer

export default expensesReducer;