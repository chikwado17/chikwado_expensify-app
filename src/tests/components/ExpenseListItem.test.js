import React from 'react';
import { shallow } from 'enzyme';
import  ExpenseListItem  from '../../components/ExpenseListItem';
import expenses from '../fixtures/expenses';



test("should test expenseListItem with expenses data", () => {
                                                //speading expenses and getting the first item
    const wrapper = shallow(<ExpenseListItem {...expenses[0]}/>);
    expect(wrapper).toMatchSnapshot();
});


