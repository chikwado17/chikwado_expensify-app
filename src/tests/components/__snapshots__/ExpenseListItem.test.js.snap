// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should test expenseListItem with expenses data 1`] = `
<div>
  <Link
    className="list-item"
    replace={false}
    to="/edit/1"
  >
    <div>
      <h3
        className="list-item__title"
      >
        uber
      </h3>
      <span
        className="list-item__sub-title"
      >
         
        January 1st, 1970
      </span>
    </div>
    <h3
      className="list-item__data"
    >
      $1.95
    </h3>
  </Link>
</div>
`;
